docs/RFCS/20210722_on_update_and_row_rehoming.md

```diff
@@ -0,0 +1,250 @@
+- Feature Name: ON UPDATE constraints and REGIONAL BY ROW rehoming
+- Status: draft
+- Start Date: 2021-07-22
+- Authors: Peyton Walters
+- RFC PR:
+- Cockroach Issue: [#28281](https://github.com/cockroachdb/cockroach/issues/28281), [#65532](https://github.com/cockroachdb/cockroach/issues/65532)
+
+# Summary
+
+This RFC proposes the addition of ON UPDATE constraints to allow the user to
+recalculate column expression on them on UPDATE/UPSERT.
+
+This RFC also proposes optionally using an ON UPDATE constraint to re-home rows
+in REGIONAL BY ROW tables on UPDATE/UPSERT, allowing for rows to be rehomed
+post-insert.
+
+# Motivation
+
+In release 21.1, we added REGIONAL BY ROW tables, allowing users to tag rows
+with a region identifier in order to home rows in the same table to different
+regions. With this addition, we added a default clause to the region column in
+case the region is not specified on INSERT, defaulting to the user’s gateway
+region and falling back to the user’s primary region in cases where the gateway
+region is not a region in the database.
+
+This design presents challenges in the case where the user performs an INSERT in
+a region that is not their “home”. For example, if an American user was to
+INSERT their data in Europe then travel back to America, their data would stay
+homed in Europe, defeating the purpose of this optimization.
+
+This design proposes the addition of an ON UPDATE constraint to allow users to
+specify an expression to be applied on UPDATE/UPSERT. This extension will then
+be used to recalculate `crdb_region` from user’s gateway region on UPDATE/UPSERT
+with the idea that continuing to home rows post-insert will lead to more
+efficient row placement.
+
+# Technical design
+
+Note: In this RFC, the term UPDATE is used to refer to behavior that should
+happen on both UPDATE and UPSERT.
+
+## Update Mechanism
+
+To support rehoming, a "recalculate on UPDATE" mechanism will be implemented.
+This mechanism will allow users to mark a column with an expression to be
+recalculated whenever a row is modified. This mechanism can be thought of as
+similar to DEFAULT expressions in that if a column is not explicitly set in an
+UPDATE, it will be filled with the value calculated from the column's ON UPDATE
+expression.
+
+Note that this ON UPDATE expression is distinct from a column's DEFAULT
+expression i.e., a user can set different DEFAULT and ON UPDATE expressions.
```

Comment:
Yes

Reply:
The behavior on INSERT is the same as if the ON UPDATE doesn't exist i.e., if the column is not specified, DEFAULT will be applied if one exists, and error otherwise.